### PR TITLE
Fixed boxed subselects with group by not compiling

### PR DESCRIPTION
**Pull request by oeed**
Fixes an issue reported on [Gitter here](https://gitter.im/diesel-rs/diesel?at=60ac7959df5c362d4bec9feb).

A query filtering on a subselect that itself uses group by, such as below, would fail to compile. Without boxing the subselect the query compiles fine.

```rust
let subselect = posts::table
  .group_by(posts::user_id)
  .select(diesel::dsl::min(posts::id))
  .into_boxed();

// get one example post for each user (for users with a post)
let source = posts::table
  .filter(posts::id.nullable().eq_any(subselect))
  .select((posts::user_id, posts::title));
```

```
error[E0277]: `BoxedSelectStatement<'_, diesel::sql_types::Nullable<diesel::sql_types::Integer>, schema::posts::table, Mysql, schema::posts::columns::user_id>` is not an iterator
   --> diesel_tests/tests/group_by.rs:230:45
    |
230 |         .filter(posts::id.nullable().eq_any(subselect))
    |                                             ^^^^^^^^^ `BoxedSelectStatement<'_, diesel::sql_types::Nullable<diesel::sql_types::Integer>, schema::posts::table, Mysql, schema::posts::columns::user_id>` is not an iterator
    |
    = help: the trait `Iterator` is not implemented for `BoxedSelectStatement<'_, diesel::sql_types::Nullable<diesel::sql_types::Integer>, schema::posts::table, Mysql, schema::posts::columns::user_id>`
    = note: required because of the requirements on the impl of `IntoIterator` for `BoxedSelectStatement<'_, diesel::sql_types::Nullable<diesel::sql_types::Integer>, schema::posts::table, Mysql, schema::posts::columns::user_id>`
    = note: required because of the requirements on the impl of `AsInExpression<diesel::sql_types::Nullable<diesel::sql_types::Integer>>` for `BoxedSelectStatement<'_, diesel::sql_types::Nullable<diesel::sql_types::Integer>, schema::posts::table, Mysql, schema::posts::columns::user_id>`
```

The issue is simply that the implementation of `AsInExpression` for `BoxedSelectStatement` omitted the `GB` generic, thus is was bound to `()` by default. This PR adds two tests, one non-boxed and another for boxed version for such a query.

### CODE REVIEW

**Comment by oeed** [comment]
Before review, I've noticed that there are 3 other references to `BoxedSelectStatement<'a, ST, QS, DB>` excluding the `GB` parameter.

```rust
// pg/expression/array_comparison.rs: 160
impl<'a, ST, QS, DB> AsArrayExpression<ST> for BoxedSelectStatement<'a, ST, QS, DB>
```

```rust
// query_builder/select_statement/boxed.rs: 381
impl<'a, ST, QS, DB> CombineDsl for BoxedSelectStatement<'a, ST, QS, DB>
```

```rust
// query_builder/upsert/into_conflict_clause.rs: 111
impl<'a, ST, QS, DB, Columns> IntoConflictValueClause
    for InsertFromSelect<BoxedSelectStatement<'a, ST, QS, DB>, Columns>
```

I haven't tried testing these, but should these also have the parameter added?

**Review by pksunkara — APPROVED** [review]
I haven't checked the other things you were talking about. Do you have failing tests for them?

**Comment by weiznich** [comment]
Thanks for submitting this pr :+1:

I've gone through the code and noticed a few more instances of this problem. Would be great to have them fixed as well.

https://github.com/diesel-rs/diesel/blob/a39c5293cbd4a4e2f677014138489c1977f07299/diesel/src/query_builder/select_statement/boxed.rs#L381 

This impl should accept a group by clause

https://github.com/diesel-rs/diesel/blob/a39c5293cbd4a4e2f677014138489c1977f07299/diesel/src/pg/expression/array_comparison.rs#L148-L149

This impl should accept a having clause

https://github.com/diesel-rs/diesel/blob/a39c5293cbd4a4e2f677014138489c1977f07299/diesel/src/pg/expression/array_comparison.rs#L160

This impl should accept a group by clause

https://github.com/diesel-rs/diesel/blob/a39c5293cbd4a4e2f677014138489c1977f07299/diesel/src/query_builder/upsert/into_conflict_clause.rs#L40-L41

This impl should accept a having clause

https://github.com/diesel-rs/diesel/blob/a39c5293cbd4a4e2f677014138489c1977f07299/diesel/src/query_builder/upsert/into_conflict_clause.rs#L51-L52

This impl should accept a group by clause

https://github.com/diesel-rs/diesel/blob/a39c5293cbd4a4e2f677014138489c1977f07299/diesel/src/query_builder/upsert/into_conflict_clause.rs#L94-L95

This impl should accept a having clause

https://github.com/diesel-rs/diesel/blob/a39c5293cbd4a4e2f677014138489c1977f07299/diesel/src/query_builder/upsert/into_conflict_clause.rs#L111-L112

This impl should accept a group by clause

https://github.com/diesel-rs/diesel/blob/a39c5293cbd4a4e2f677014138489c1977f07299/diesel/src/expression/array_comparison.rs#L123

This impl should accept a having clause

**Comment by oeed** [comment]
Great, will go through and fix those this week when I have the opportunity.

File: diesel/src/pg/expression/array_comparison.rs [hunk 1]
```diff
@@ -145,8 +145,8 @@ where
     }
 }
 
-impl<ST, S, F, W, O, L, Of, G, FU> AsArrayExpression<ST>
-    for SelectStatement<S, F, W, O, L, Of, G, FU>
+impl<ST, F, S, D, W, O, LOf, G, H, LC> AsArrayExpression<ST>
```
Comment (oeed) [inline diesel/src/pg/expression/array_comparison.rs:148]: This occurrence seemed to be entirely incorrect with its ordering. Have amended but can revert if this was intentional.

File: diesel/src/expression/array_comparison.rs [hunk 2]
```diff
@@ -120,7 +120,8 @@ pub trait MaybeEmpty {
     fn is_empty(&self) -> bool;
 }
 
-impl<ST, S, F, W, O, L, Of, G, LC> AsInExpression<ST> for SelectStatement<S, F, W, O, L, Of, G, LC>
+impl<ST, F, S, D, W, O, LOf, G, H, LC> AsInExpression<ST>
```
Comment (oeed) [inline diesel/src/expression/array_comparison.rs:123]: Same applies with incorrect ordering here.

**Comment by oeed** [comment]
Apologies this took longer than the week stated, had other pressing issues come up.

All mentioned occurrences are fixed, just worth checking out the comments above.

**Comment by weiznich** [comment]
Thanks for the update. Looks good now :+1:

**Pull request merged**